reset.c

```diff
@@ -0,0 +1,849 @@
+
+    if (!infile || !outfile) {
+        printf("Invalid parameters in reset!\n");
+        return EXIT_FAILURE;
+    }
+
+    //read input header
+    in_samhdr = sam_hdr_read(infile);
+    if (!in_samhdr)
+    {
+        printf("Failed to read header from file\n");
+        return EXIT_FAILURE;
+    }
+    //create output header
+    if (!(out_samhdr = sam_hdr_init()))
+    {
+        printf("Failed to create output header!\n");
+        sam_hdr_destroy(in_samhdr);
+        return EXIT_FAILURE;
+    }
+    //TODO SO/GO status not maintained, do they hold as secondary n supplementary are removed?
+    //add version to output header
+    if  (-1 == sam_hdr_add_line(out_samhdr,"HD", "VN", SAM_FORMAT_VERSION, NULL)) {
+        printf("Failed to set header data (%d)\n", errno);
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+    //add RG / PG lines if configured
+    if ((config->keepRGs && getRGlines(in_samhdr, out_samhdr)) ||
+            getPGlines(in_samhdr, out_samhdr, config, args)) {
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+
+    //write output header
+    if (sam_hdr_write(outfile, out_samhdr)) {
+        printf("Failed to write output header (%d)\n", errno);
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        return EXIT_FAILURE;
+    }
+
+    int ret_r = 0, ret_w = 0;
+    bam1_t *bamdata = bam_init1();      //input bam
+    bam1_t *outdata = bam_init1();      //output bam
+    if (!bamdata || !outdata)
+    {
+        printf("Failed to allocate data memory!\n");
+        sam_hdr_destroy(in_samhdr);
+        sam_hdr_destroy(out_samhdr);
+        if (!bamdata)
+            bam_destroy1(bamdata);
+        if (!outdata)
+            bam_destroy1(outdata);
+        return EXIT_FAILURE;
+    }
+
+    errno = 0;
+    int cnt = 0, i = 0;
+    kstring_t querydata = KS_INITIALIZE, qualdata = KS_INITIALIZE;
+    //get bam data, make updates and dump to output
+    while (0 <= (ret_r = sam_read1(infile, in_samhdr, bamdata)))
+    {
+        // read data
+        cnt++;
+        if (bamdata->core.flag & BAM_FSECONDARY || bamdata->core.flag & BAM_FSUPPLEMENTARY) {
+            //printf("Skipping secondary/supplimentary alignment (%d %s)\n", bamdata->id, bam_get_qname(bamdata));
+            continue;
+        }
+
+        //update flags
+        uint16_t flags = bamdata->core.flag & ~BAM_FPROPER_PAIR;    //reset pair info
+        flags |= BAM_FUNMAP | BAM_FMUNMAP;                          //mark as unmapped
+        flags &= ~BAM_FMREVERSE;                                    //reset mate orientation
+
+        ks_resize(&querydata, bamdata->core.l_qseq);                //get enough space
+        ks_clear(&querydata);
+
+        ks_resize(&qualdata, bamdata->core.l_qseq);                 //get enough space
+        ks_clear(&qualdata);
```

Comment:
Both `querydata` and `qualdata` are a small memory leak.

Reply:
Added ks_free